NotificationService/NotificationService.swift

```diff
@@ -0,0 +1,151 @@
+//
+//  NotificationService.swift
+//  NotificationService
+//
+//  Created by Anton Wyrowski on 13.12.22.
+//
+
+import UserNotifications
+
+enum RSAKeyPairError: Error {
+    case failedObtainingPrivateKeyFromKeyChain
+    case failedObtainingStringRepresentationOfPrivateKey
+}
+
+
+class NotificationService: UNNotificationServiceExtension {
+
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var bestAttemptContent: UNMutableNotificationContent?
+    
+    private static let privateKeyApplicationTag = "de.tum.tca.keys.push_rsa_key"
+    private static let keychainAccessGroupName = "2J3C6P6X3N.de.tum.tca.notificationextension"
+    
+    private var privateKeyKeychainQuery: [String: Any] = [
+                kSecClass as String: kSecClassKey,
+                kSecAttrKeyClass as String: kSecAttrKeyClassPrivate,
+                kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
+                kSecAttrKeySizeInBits as String: 2048,
+                kSecAttrApplicationTag as String: privateKeyApplicationTag,
+                kSecReturnRef as String: true,
+                kSecAttrAccessGroup as String: NotificationService.keychainAccessGroupName
+            ]
+
+    /**
+     Called before the push notification is displayed to the user
+     */
+    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
+        self.contentHandler = contentHandler
+        bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
+        
+        if let bestAttemptContent = bestAttemptContent {
+            
+            let exists = checkIfPrivateKeyAlreadyExists()
+            
+            func shouldNotHappen(_ error: String = "") {
```

Comment:
Bit ugly, think about restructuring

Reply:
You mean the `shouldNotHappenFunction` ?

Reply:
Yes, function inside a function...